### PR TITLE
Fix crash when posting notification to notification center from dealloc

### LINKED ISSUE #577: Posting NSNotification during dealloc causes crash.
If an object posts a notification during dealloc passing `self` to the `object` argument it causes a crash.

This might seem an odd thing to do :-) but it turns out some of our code during dealloc and cleanup generates a few model change notifications, so thats why we experience this crash.

Example code: 

```
@implementation TestObjectA

- (instancetype)init
{
	if (self = [super init]) {
	}

	return self;
}

- (void)dealloc
{
	[[NSNotificationCenter defaultCenter] postNotificationName:@"MOOSENOTIFICATION" object:self];
}

@end
```

```
TestObjectA *object = [[TestObjectA alloc] init];

// Let ARC dealloc object on out of scope
```

### PR DESCRIPTION
**Pull request by rajsesh-msft**
NSNotificationCenter was using an auto released NSNotification object, which was in turn holding a reference to an object that is going away.  This change removes the usage of autorelease to make the api compatible with reference platform.

Fixes #577

### CODE REVIEW

**Comment by rajsesh-msft** [comment]
@bbowman is added to the review. #Closed

**Comment by rajsesh-msft** [comment]
@mnithish is added to the review. #Closed

File: Frameworks/Foundation/NSNotification.mm [hunk 1]
```diff
@@ -85,9 +73,13 @@ - (void)dealloc {
  @Status Stub
  @Notes
 */
-- (instancetype)initWithName:(NSString*)aName object:(id)object userInfo:(NSDictionary*)userInfo {
-    UNIMPLEMENTED();
-    return StubReturn();
+- (instancetype)initWithName:(NSString*)aName object:(id)object userInfo:(NSDictionary*)info {
+    if ((self = [super init]) != nil) {
```
Comment (msft-Jeyaram) [inline Frameworks/Foundation/NSNotification.mm]: nit: not needed

**Comment by msft-Jeyaram** [comment]
>     id userInfo;

nit: could we just wrap them in StrongId? 

---
Refers to: Frameworks/Foundation/NSNotification.mm:23 in 9291c4e. [](commit_id = 9291c4ea7d40c429dd9db7728587d1b95fe07fee, deletion_comment = False)

File: Frameworks/Foundation/NSNotification.mm [hunk 2]
```diff
@@ -74,8 +62,8 @@ - (NSDictionary*)userInfo {
  @Status Interoperable
 */
 - (void)dealloc {
-    [notificationName release];
     [notificationObj release];
+    [notificationName release];
```
Comment (mnithish) [inline Frameworks/Foundation/NSNotification.mm]: Recommend using StrongId<> so these can be removed

**Comment by mnithish** [comment]
:shipit:

**Comment by msft-Jeyaram** [comment]
:shipit:

File: Frameworks/Foundation/NSNotification.mm [hunk 3]
```diff
@@ -85,9 +73,13 @@ - (void)dealloc {
...
  @Notes
 */
-- (instancetype)initWithName:(NSString*)aName object:(id)object userInfo:(NSDictionary*)userInfo {
-    UNIMPLEMENTED();
-    return StubReturn();
+- (instancetype)initWithName:(NSString*)aName object:(id)object userInfo:(NSDictionary*)info {
+    if ((self = [super init]) != nil) {
+        notificationObj = [object retain];
+        notificationName = [aName copy];
+        userInfo = [userInfo retain];
```
Comment (rajsesh-msft) [inline Frameworks/Foundation/NSNotification.mm]: bug bug!

**Comment by rajsesh-msft** [comment]
>     id userInfo;

I'd like to go in the other direction instead and just use the storage provided by the property declaration. removes a bunch of unnecessary code as well.

---
In reply to: [229817516](https://github.com/Microsoft/WinObjC/pull/600#issuecomment-229817516) [](ancestors = 229817516)

---
Refers to: Frameworks/Foundation/NSNotification.mm:23 in 9291c4e. [](commit_id = 9291c4ea7d40c429dd9db7728587d1b95fe07fee, deletion_comment = False)

**Comment by mnithish** [comment]
>     id userInfo;

That sounds like a better idea. Let property do the memory management

---
In reply to: [229832711](https://github.com/Microsoft/WinObjC/pull/600#issuecomment-229832711) [](ancestors = 229832711,229817516)

---
Refers to: Frameworks/Foundation/NSNotification.mm:23 in 9291c4e. [](commit_id = 9291c4ea7d40c429dd9db7728587d1b95fe07fee, deletion_comment = False)

File: Frameworks/Foundation/NSNotification.mm (hunk 3 above)
Comment (msft-Jeyaram) [inline Frameworks/Foundation/NSNotification.mm]: info :D

---
In reply to: [69236021](https://github.com/Microsoft/WinObjC/pull/600#discussion_r69236021) [](ancestors = 69236021)

**Comment by msft-Jeyaram** [comment]
>     id userInfo;

although you can still copy it and attach to it (so you aren't getting a retain), at the end it would give the auto-autorelease mechanism.
Given we have these properties backed by internal ivars, we could ease the pain and use StrongId (which comes with it's own shenanigans). 

Your direction makes sense, we can get rid of the these and directly set the properties. 

---
In reply to: [229856697](https://github.com/Microsoft/WinObjC/pull/600#issuecomment-229856697) [](ancestors = 229856697,229832711,229817516)

---
Refers to: Frameworks/Foundation/NSNotification.mm:23 in 9291c4e. [](commit_id = 9291c4ea7d40c429dd9db7728587d1b95fe07fee, deletion_comment = False)

File: Frameworks/Foundation/NSNotificationCenter.mm [hunk 4]
```diff
@@ -79,8 +79,9 @@ - (instancetype)init {
  @Status Interoperable
 */
 - (void)postNotificationName:(NSString*)name object:(NSObject*)sender {
-    NSNotification* notification = [NSNotification notificationWithName:name object:sender];
+    NSNotification* notification = [[NSNotification alloc] initWithName:name object:sender userInfo: nil];
```
Comment (msft-Jeyaram) [inline Frameworks/Foundation/NSNotificationCenter.mm]: format issue

File: Frameworks/Foundation/NSNotification.mm [hunk 5]
```diff
@@ -85,9 +48,13 @@ - (void)dealloc {
...
  @Notes
 */
-- (instancetype)initWithName:(NSString*)aName object:(id)object userInfo:(NSDictionary*)userInfo {
-    UNIMPLEMENTED();
-    return StubReturn();
+- (instancetype)initWithName:(NSString*)aName object:(id)object userInfo:(NSDictionary*)info {
+    if ((self = [super init]) != nil) {
+        _object = [object retain];
+        _name = [aName copy];
+        _userInfo = [info retain];
```
Comment (mnithish) [inline Frameworks/Foundation/NSNotification.mm]: Should be copy

File: Frameworks/Foundation/NSNotification.mm [hunk 6]
```diff
@@ -85,9 +48,13 @@ - (void)dealloc {
  @Status Stub
```
Comment (MSFTFox) [inline Frameworks/Foundation/NSNotification.mm]: Is this still a stubbed method?

File: tests/unittests/Foundation/NSNotificationTests.mm [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+@implementation TestObjectA
+- (void)dealloc {
+    [[NSNotificationCenter defaultCenter] postNotificationName:@"MOOSENOTIFICATION" object:self];
+    [super dealloc];
+}
+
+@end
+
+TEST(NSNotificationCenter, PostNotificationFromDealloc) {
+    TestObjectA* obj = [[TestObjectA new] autorelease];
```
Comment (mnithish) [inline tests/unittests/Foundation/NSNotificationTests.mm]: nit: do you want to perform this test synchronously? i.e. allocated and free NSNotificationCenter within the test and not catch it as part of the test clean-up

Comment (rajsesh-msft) [inline tests/unittests/Foundation/NSNotificationTests.mm]: I have tested it with a release, but this is the test code from the issue report, so i decided to leave the autorelease version in.

---
In reply to: [69334437](https://github.com/Microsoft/WinObjC/pull/600#discussion_r69334437) [](ancestors = 69334437)

**Pull request merged**